src/libxtp/gwbse/bse.cc

```diff
@@ -33,62 +41,195 @@ void BSE::SetupDirectInteractionOperator() {
-  _epsilon_0_inv = VectorXfd::Zero(es.eigenvalues().size());
+
+  _epsilon_0_inv = Eigen::VectorXd::Zero(es.eigenvalues().size());
   for (int i = 0; i < es.eigenvalues().size(); ++i) {
     if (es.eigenvalues()(i) > 1e-8) {
       _epsilon_0_inv(i) = 1 / es.eigenvalues()(i);
     }
   }
 }
 
+template <typename BSE_OPERATOR>
+void BSE::configureBSEOperator(BSE_OPERATOR& H) {
+  BSEOperator_Options opt;
+  opt.cmax = _opt.cmax;
+  opt.homo = _opt.homo;
+  opt.qpmin = _opt.qpmin;
+  opt.rpamin = _opt.rpamin;
+  opt.vmin = _opt.vmin;
+  H.configure(opt);
+}
+
 void BSE::Solve_triplets() {
-  MatrixXfd H = MatrixXfd::Zero(_bse_size, _bse_size);
-  Add_Hd<real_gwbse>(H);
-  Add_Hqp<real_gwbse>(H);
+
+  TripletOperator_TDA Ht(_epsilon_0_inv, _log, _Mmn, _Hqp);
+  configureBSEOperator(Ht);
   CTP_LOG(ctp::logDEBUG, _log)
       << ctp::TimeStamp() << " Setup TDA triplet hamiltonian " << flush;
-  CTP_LOG(ctp::logDEBUG, _log) << ctp::TimeStamp() << " Solving for first "
-                               << _opt.nmax << " eigenvectors" << flush;
-  tools::linalg_eigenvalues(H, _bse_triplet_energies, _bse_triplet_coefficients,
-                            _opt.nmax);
+
+  solve_hermitian(Ht, _bse_triplet_energies, _bse_triplet_coefficients);
+
   return;
 }
 
 void BSE::Solve_singlets() {
   if (_opt.useTDA) {
     Solve_singlets_TDA();
   } else {
+    if (_opt.davidson) {
+      CTP_LOG(ctp::logDEBUG, _log)
+          << ctp::TimeStamp()
+          << " Davidson solver not implemented for BTDA. Full diagonalization."
+          << flush;
+      _opt.davidson = 0;
+    }
     Solve_singlets_BTDA();
   }
 }
 
 void BSE::Solve_singlets_TDA() {
-  MatrixXfd H = MatrixXfd::Zero(_bse_size, _bse_size);
-  Add_Hd<real_gwbse>(H);
-  Add_Hqp<real_gwbse>(H);
-  Add_Hx<real_gwbse, 2>(H);
+
+  SingletOperator_TDA Hs(_epsilon_0_inv, _log, _Mmn, _Hqp);
+  configureBSEOperator(Hs);
   CTP_LOG(ctp::logDEBUG, _log)
       << ctp::TimeStamp() << " Setup TDA singlet hamiltonian " << flush;
-  CTP_LOG(ctp::logDEBUG, _log) << ctp::TimeStamp() << " Solving for first "
-                               << _opt.nmax << " eigenvectors" << flush;
-  tools::linalg_eigenvalues(H, _bse_singlet_energies, _bse_singlet_coefficients,
-                            _opt.nmax);
-  return;
+
+  solve_hermitian(Hs, _bse_singlet_energies, _bse_singlet_coefficients);
 }
 
 void BSE::SetupHs() {
-  _eh_s = MatrixXfd::Zero(_bse_size, _bse_size);
-  Add_Hd<real_gwbse>(_eh_s);
-  Add_Hqp<real_gwbse>(_eh_s);
-  Add_Hx<real_gwbse, 2>(_eh_s);
+
+  SingletOperator_TDA Hs(_epsilon_0_inv, _log, _Mmn, _Hqp);
+  configureBSEOperator(Hs);
+  _eh_s = Hs.get_full_matrix();
 }
 
 void BSE::SetupHt() {
-  _eh_t = MatrixXfd::Zero(_bse_size, _bse_size);
-  Add_Hd<real_gwbse>(_eh_t);
-  Add_Hqp<real_gwbse>(_eh_t);
+
+  TripletOperator_TDA Ht(_epsilon_0_inv, _log, _Mmn, _Hqp);
+  configureBSEOperator(Ht);
+  _eh_t = Ht.get_full_matrix();
+}
+
+Eigen::MatrixXd BSE::GetComponentMatrix(std::string name) {
```

Comment:
Is this only for testing? If so, please remove and put it into the test. We do not add functions just for testing

Reply:
indeed it's better. I kept it as a method of the `BSE` class though ... which means I still have the prototype in `bse.h`. Is that OK  ?

Reply:
No. My opinion is sort of if you need a method just for testing put it in the test.

Reply:
When someone wants to use the bse class, (s)he should not find this function.

Reply:
Yeah I agree. The issue is that we have to access a few members that are private to `BSE` (_log, _Mmn, ... )I'll find a way around

Reply:
Made a derived class in the test and made those members protected in `bse`. Quite a lot of trouble not to simply make them public !! :)

Reply:
Why make the class in the first place, you call the function once for each argument. Just copy the code to the test and delete the class.

Reply:
I changed it